javascript/features/death_match/death_match_commands.js

```diff
@@ -3,27 +3,55 @@
 // be found in the LICENSE file.
 
 import { CommandBuilder } from 'components/command_manager/command_builder.js';
+import { DeathMatchLocation } from 'features/death_match/death_match_location.js';
+import { Menu } from 'components/menu/menu.js';
 
 // Contains the /dm command for players to allow to teleport to a DM zone with specific weapons.
 export class DeathMatchCommands {    
 
-    constructor(manager) {
+    constructor(abuse, manager) {
         server.commandManager.buildCommand('deathmatch')
-            .parameters([{ name: 'zone', type: CommandBuilder.NUMBER_PARAMETER }])
+            .parameters([{ name: 'zone', type: CommandBuilder.NUMBER_PARAMETER, optional: true }])
             .sub('leave')
                 .build(DeathMatchCommands.prototype.onLeaveCommand.bind(this))
             .sub('stats')
                 .build(DeathMatchCommands.prototype.onStatsCommand.bind(this))
     
         .build(DeathMatchCommands.prototype.onDmCommand.bind(this));
 
-        this.manager_ = manager;
+        this.manager_ = manager;        
+        this.abuse_ = abuse;
     }
 
     // The death match command is being used.
-    onDmCommand(player, zone) { 
-        if(zone === null || zone === undefined || typeof zone !== 'number') {
-            player.sendMessage(Message.DEATH_MATCH_INVALID_ZONE, zone);
+    async onDmCommand(player, zone) { 
+        if(isNaN(zone) || !DeathMatchLocation.hasLocation(zone)) {
+            const dialog = new Menu('Choose a death match zone.', ['Zone', 'Name', 'Teams', 'Lag shot']);
+            for(const zoneId of this.manager_.validDmZones()) {
+                const location = DeathMatchLocation.getById(zoneId);  
+                dialog.addItem(location.id, location.name, location.hasTeams ? "Yes" : "No", 
+                                location.lagShot ? "Yes": "No", 
+                    (player) => {
+                        this.manager_.goToDmZone(player, zoneId);
```

Comment:
This is now abusable: you can `/deathmatch` and then quickly press enter (to accept the first zone in the dialog) to enter a deathmatch zone without the activity check on line :43, or the teleportation check on line :52.

I would suggest splitting this up in two methods:

```javascript
async onDmCommand(player, zone) {
    if (DeathMatchLocations.hasLocation(zone))
        return enterDeathMatchZone(zone, player);

    // .. show dialog ..
    dialog.addItem(
        'foo', 'bar', 'baz',
        DeathMatchCommands.prototype.enterDeathMatchZone.bind(this, zone));
}

async enterDeathMatchZone(zone, player) {
    if (player.activity …) {}
    if (teleportStatus…) {}

    this.manager_.goToDmZone(player, zone);
}
```

Why reverse the argument order for `enterDeathMatchZone`? Because it means that you can use the `bind` function without having to create a listener, like you do now 😃